File: uva/v13/p1330-City-Game.py
#! /usr/bin/python3
# 3 star dp
# Runtime error.

import sys

k = int(input().strip())
for inputLine in sys.stdin:
    if inputLine.strip() == '':
        continue
    m, n = list(map(int, inputLine.strip().split(' ')))
    maps = [['R' for i in range(n + 1)]]
    for x in range(m):
        row = input().strip().split(' ')
        row.insert(0, 'R')
        maps.append(row)

    width = [[0 for j in range(n + 1)] for i in range(m + 1)]
    for x in range(1, m + 1):
        for y in range(1, n + 1):
            if maps[x][y] == 'F':
                width[x][y] = width[x][y - 1] + 1

    maxArea = 0
    for x in range(1, m + 1):
        for y in range(1, n + 1):
            if maps[x][y] == 'F':
                maxWidth = width[x][y]
                for z in range(x, 0, -1):
                    if maps[z][y] == 'R':
                        break
                    maxWidth = min(maxWidth, width[z][y])
                    area = maxWidth * (x - z + 1)
                    if maxArea < area:
                        maxArea = area
    print(maxArea * 3)

File: uva/v13/p1330-City-Game.cpp
/*
 * 3 star
 * DP
 * 24420565	1330	City Game	Accepted	C++11	0.080	2020-01-14 09:07:05
 */
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
	int t;
	cin >> t;
	while (t--) {
		int m, n;
		cin >> m >> n;
		string cell;

		// create a map with wall at top and left
		vector<vector<char> > maps(m + 1 , vector<char> (n + 1, 'R'));
		for (int i = 1; i <= m; ++i) {
			for (int j = 1; j <= n; ++j) {
				cin >> cell;
				maps[i][j] = cell[0];
			}
		}

		vector<vector<int> > width(m + 1, vector<int> (n + 1, 0));
		for (int i = 1; i <= m; ++i) {
			for (int j = 1; j <= n; ++j) {
				if (maps[i][j] == 'F') {
					width[i][j] = width[i][j - 1] + 1;
				}
			}
		}

		int maxArea = 0;
		for (int x = 1; x <= m; ++x) {
			for (int y = 1; y <= n; ++y) {
				int maxWidth = width[x][y];
				for (int z = x; maps[z][y] == 'F'; --z) {
					maxWidth = min(width[z][y], maxWidth);
					int Area = maxWidth * (x - z + 1);
					if (maxArea < Area) {
						maxArea = Area;
					}
				}
			}
		}
		cout << maxArea * 3 << endl;
	}
}
